src/nvapi_d3d12.cpp

```diff
@@ -272,14 +272,21 @@ extern "C" {
     }
 
     static bool ConvertBuildRaytracingAccelerationStructureInputs(const NVAPI_D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_INPUTS_EX* nvDesc, std::vector<D3D12_RAYTRACING_GEOMETRY_DESC>& geometryDescs, D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_INPUTS* d3dDesc) {
-        *d3dDesc = {
-            .Type = nvDesc->type,
-            // assume that OMM via VK_EXT_opacity_micromap is not supported, allow only standard flags to be passed
-            .Flags = static_cast<D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAGS>(nvDesc->flags & 0x3f),
-            .NumDescs = nvDesc->numDescs,
-            .DescsLayout = nvDesc->descsLayout,
-            .InstanceDescs = nvDesc->instanceDescs,
-        };
+        d3dDesc->Type = nvDesc->type;
+        // assume that OMM via VK_EXT_opacity_micromap is not supported, allow only standard flags to be passed
+        d3dDesc->Flags = static_cast<D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAGS>(nvDesc->flags & 0x3f);
+        d3dDesc->NumDescs = nvDesc->numDescs;
+        d3dDesc->DescsLayout = nvDesc->descsLayout;
+
+        if (d3dDesc->Type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_TOP_LEVEL) {
+            d3dDesc->InstanceDescs = nvDesc->instanceDescs;
+            return true;
+        }
+
+        if (d3dDesc->Type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && d3dDesc->DescsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY_OF_POINTERS) {
+            d3dDesc->ppGeometryDescs = reinterpret_cast<const D3D12_RAYTRACING_GEOMETRY_DESC* const*>(nvDesc->ppGeometryDescs);
+            return true;
+        }
 
         if (d3dDesc->Type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && d3dDesc->DescsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY) {
             if (nvDesc->geometryDescStrideInBytes != sizeof(NVAPI_D3D12_RAYTRACING_GEOMETRY_DESC_EX))
```

Comment:
As stated on chat, from what I saw, CP2077 never uses this path, it always uses array-of-pointers. An alternative might be to drop the code and return no-implementation for this scenario. We could resurrect the code once we have a title that uses this path and can thus ensure that the implementation is correct, eg the lifetime of the vector we are creating.

Reply:
Doing so would result in something like this:

<details>
  <summary>Code</summary>

```c++
    NvAPI_Status __cdecl NvAPI_D3D12_GetRaytracingAccelerationStructurePrebuildInfoEx(ID3D12Device5* pDevice, NVAPI_GET_RAYTRACING_ACCELERATION_STRUCTURE_PREBUILD_INFO_EX_PARAMS* pParams) {
        constexpr auto n = __func__;
        static bool alreadyLoggedOk = false;

        if (pDevice == nullptr || pParams == nullptr)
            return InvalidArgument(n);

        if (pParams->version != NVAPI_GET_RAYTRACING_ACCELERATION_STRUCTURE_PREBUILD_INFO_EX_PARAMS_VER1)
            return IncompatibleStructVersion(n);

        if (pParams->pDesc == nullptr || pParams->pInfo == nullptr)
            return InvalidArgument(n);

        if (pParams->pDesc->type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && pParams->pDesc->descsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY)
            return NoImplementation(n);

        D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_INPUTS desc = {
            .Type = pParams->pDesc->type,
            // assume that OMM via VK_EXT_opacity_micromap is not supported, allow only standard flags to be passed
            .Flags = static_cast<D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAGS>(pParams->pDesc->flags & 0x3f),
            .NumDescs = pParams->pDesc->numDescs,
            .DescsLayout = pParams->pDesc->descsLayout,
            .InstanceDescs = {}};

        if (pParams->pDesc->type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_TOP_LEVEL)
            desc.InstanceDescs = pParams->pDesc->instanceDescs;

        if (pParams->pDesc->type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && pParams->pDesc->descsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY_OF_POINTERS)
            desc.ppGeometryDescs = reinterpret_cast<const D3D12_RAYTRACING_GEOMETRY_DESC* const*>(pParams->pDesc->ppGeometryDescs);

        pDevice->GetRaytracingAccelerationStructurePrebuildInfo(&desc, pParams->pInfo);

        return Ok(n, alreadyLoggedOk);
    }

    NvAPI_Status __cdecl NvAPI_D3D12_BuildRaytracingAccelerationStructureEx(ID3D12GraphicsCommandList4* pCommandList, const NVAPI_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_EX_PARAMS* pParams) {
        constexpr auto n = __func__;
        static bool alreadyLoggedOk = false;

        if (pCommandList == nullptr || pParams == nullptr)
            return InvalidArgument(n);

        if (pParams->version != NVAPI_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_EX_PARAMS_VER1)
            return IncompatibleStructVersion(n);

        if (pParams->pDesc == nullptr || (pParams->numPostbuildInfoDescs != 0 && pParams->pPostbuildInfoDescs == nullptr))
            return InvalidArgument(n);

        if (pParams->pDesc->inputs.type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && pParams->pDesc->inputs.descsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY)
            return NoImplementation(n);

        std::vector<D3D12_RAYTRACING_GEOMETRY_DESC> geometryDescs{};
        D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_DESC desc = {
            .DestAccelerationStructureData = pParams->pDesc->destAccelerationStructureData,
            .Inputs = {
                .Type = pParams->pDesc->inputs.type,
                // assume that OMM via VK_EXT_opacity_micromap is not supported, allow only standard flags to be passed
                .Flags = static_cast<D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAGS>(pParams->pDesc->inputs.flags & 0x3f),
                .NumDescs = pParams->pDesc->inputs.numDescs,
                .DescsLayout = pParams->pDesc->inputs.descsLayout,
                .InstanceDescs = {}},
            .SourceAccelerationStructureData = pParams->pDesc->sourceAccelerationStructureData,
            .ScratchAccelerationStructureData = pParams->pDesc->scratchAccelerationStructureData,
        };

        if (pParams->pDesc->inputs.type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_TOP_LEVEL)
            desc.Inputs.InstanceDescs = pParams->pDesc->inputs.instanceDescs;

        if (pParams->pDesc->inputs.type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && pParams->pDesc->inputs.descsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY_OF_POINTERS)
            desc.Inputs.ppGeometryDescs = reinterpret_cast<const D3D12_RAYTRACING_GEOMETRY_DESC* const*>(pParams->pDesc->inputs.ppGeometryDescs);

        pCommandList->BuildRaytracingAccelerationStructure(&desc, pParams->numPostbuildInfoDescs, pParams->pPostbuildInfoDescs);

        return Ok(n, alreadyLoggedOk);
    }
``` 

</details>

Tbh, I start to favor this dumbed down version. CP2077 works with this version (I've tested that) and the array implementation is preserved in git until we need to resurrect it.

Reply:
Thinking more about this, may be we should just forward everything, but bail out when NV-input type has been specified (currently only `NVAPI_D3D12_RAYTRACING_GEOMETRY_TYPE_OMM_TRIANGLES_EX`):

<details>
  <summary>Code</summary>

```c++
NvAPI_Status __cdecl NvAPI_D3D12_GetRaytracingAccelerationStructurePrebuildInfoEx(ID3D12Device5* pDevice, NVAPI_GET_RAYTRACING_ACCELERATION_STRUCTURE_PREBUILD_INFO_EX_PARAMS* pParams) {
        constexpr auto n = __func__;
        static bool alreadyLoggedOk = false;
        static bool alreadyLoggedNoImplementation = false;

        if (pDevice == nullptr || pParams == nullptr)
            return InvalidArgument(n);

        if (pParams->version != NVAPI_GET_RAYTRACING_ACCELERATION_STRUCTURE_PREBUILD_INFO_EX_PARAMS_VER1)
            return IncompatibleStructVersion(n);

        if (pParams->pDesc == nullptr || pParams->pInfo == nullptr)
            return InvalidArgument(n);

        for (auto i = 0U; i < pParams->pDesc->numDescs; i++)
            if (pParams->pDesc->pGeometryDescs[i].type == NVAPI_D3D12_RAYTRACING_GEOMETRY_TYPE_OMM_TRIANGLES_EX)
                return NoImplementation(n, alreadyLoggedNoImplementation); // GetRaytracingCaps reports no OMM caps, so hopefully no one attempts to try

        D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_INPUTS desc{};
        desc.Type = pParams->pDesc->type;
        // assume that OMM via VK_EXT_opacity_micromap is not supported, allow only standard flags to be passed
        desc.Flags = static_cast<D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAGS>(pParams->pDesc->flags & 0x3f);
        desc.NumDescs = pParams->pDesc->numDescs;
        desc.DescsLayout = pParams->pDesc->descsLayout;

        if (pParams->pDesc->type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_TOP_LEVEL)
            desc.InstanceDescs = pParams->pDesc->instanceDescs;

        if (pParams->pDesc->type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && pParams->pDesc->descsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY_OF_POINTERS)
            desc.pGeometryDescs = reinterpret_cast<const D3D12_RAYTRACING_GEOMETRY_DESC*>(pParams->pDesc->pGeometryDescs);

        if (pParams->pDesc->type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && pParams->pDesc->descsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY_OF_POINTERS)
            desc.ppGeometryDescs = reinterpret_cast<const D3D12_RAYTRACING_GEOMETRY_DESC* const*>(pParams->pDesc->ppGeometryDescs);

        pDevice->GetRaytracingAccelerationStructurePrebuildInfo(&desc, pParams->pInfo);

        return Ok(n, alreadyLoggedOk);
    }

    NvAPI_Status __cdecl NvAPI_D3D12_BuildRaytracingAccelerationStructureEx(ID3D12GraphicsCommandList4* pCommandList, const NVAPI_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_EX_PARAMS* pParams) {
        constexpr auto n = __func__;
        static bool alreadyLoggedOk = false;
        static bool alreadyLoggedNoImplementation = false;

        if (pCommandList == nullptr || pParams == nullptr)
            return InvalidArgument(n);

        if (pParams->version != NVAPI_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_EX_PARAMS_VER1)
            return IncompatibleStructVersion(n);

        if (pParams->pDesc == nullptr || (pParams->numPostbuildInfoDescs != 0 && pParams->pPostbuildInfoDescs == nullptr))
            return InvalidArgument(n);

        for (auto i = 0U; i < pParams->pDesc->inputs.numDescs; i++)
            if (pParams->pDesc->inputs.pGeometryDescs[i].type == NVAPI_D3D12_RAYTRACING_GEOMETRY_TYPE_OMM_TRIANGLES_EX)
                return NoImplementation(n, alreadyLoggedNoImplementation); // GetRaytracingCaps reports no OMM caps, so hopefully no one attempts to try

        std::vector<D3D12_RAYTRACING_GEOMETRY_DESC> geometryDescs{};
        D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_DESC desc = {
            .DestAccelerationStructureData = pParams->pDesc->destAccelerationStructureData,
            .Inputs = {},
            .SourceAccelerationStructureData = pParams->pDesc->sourceAccelerationStructureData,
            .ScratchAccelerationStructureData = pParams->pDesc->scratchAccelerationStructureData,
        };
        desc.Inputs.Type = pParams->pDesc->inputs.type;
        // assume that OMM via VK_EXT_opacity_micromap is not supported, allow only standard flags to be passed
        desc.Inputs.Flags = static_cast<D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAGS>(pParams->pDesc->inputs.flags & 0x3f);
        desc.Inputs.NumDescs = pParams->pDesc->inputs.numDescs;
        desc.Inputs.DescsLayout = pParams->pDesc->inputs.descsLayout;

        if (pParams->pDesc->inputs.type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_TOP_LEVEL)
            desc.Inputs.InstanceDescs = pParams->pDesc->inputs.instanceDescs;

        if (pParams->pDesc->inputs.type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && pParams->pDesc->inputs.descsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY_OF_POINTERS)
            desc.Inputs.pGeometryDescs = reinterpret_cast<const D3D12_RAYTRACING_GEOMETRY_DESC*>(pParams->pDesc->inputs.pGeometryDescs);

        if (pParams->pDesc->inputs.type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && pParams->pDesc->inputs.descsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY_OF_POINTERS)
            desc.Inputs.ppGeometryDescs = reinterpret_cast<const D3D12_RAYTRACING_GEOMETRY_DESC* const*>(pParams->pDesc->inputs.ppGeometryDescs);

        pCommandList->BuildRaytracingAccelerationStructure(&desc, pParams->numPostbuildInfoDescs, pParams->pPostbuildInfoDescs);

        return Ok(n, alreadyLoggedOk);
    }
```

</details>

Reply:
ah yikes, CP2077 still uses `NVAPI_D3D12_RAYTRACING_GEOMETRY_TYPE_OMM_TRIANGLES_EX`, so the check for that prevents CP2077 to start RTX :(. Weird though, since it tests both capabilities with `NvAPI_D3D12_GetRaytracingCaps` when starting.

Reply:
> An alternative might be to drop the code and return no-implementation for this scenario.

Sure, that's also fine, but I'm kinda afraid CP2077 or some other game will decide to use that version in the future and then it's just more work for us to restore that. I'd prefer to have an implementation that fully supports whatever any game could throw our way as long as it correctly avoids using any OMM stuff.

> ensure that the implementation is correct, eg the lifetime of the vector we are creating

Wrt that, I checked the docs for both `BuildRaytracingAccelerationStructure` and `vkCmdBuildAccelerationStructuresKHR` and neither seem to require we keep that memory alive after the call returns. I also tried to ask around and Rhedox is pretty sure we can just free this memory immediately afterwards.

> Also considering https://microsoft.github.io/DirectX-Specs/d3d/Raytracing.html#d3d12_build_raytracing_acceleration_structure_inputs , I don't think we should treat `pGeometryDescs` and `ppGeometryDescs` differently.

I'm not exactly sure what do you mean by this, but…

```cpp
if (pParams->pDesc->inputs.type == D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_BOTTOM_LEVEL && pParams->pDesc->inputs.descsLayout == D3D12_ELEMENTS_LAYOUT_ARRAY)
    desc.Inputs.pGeometryDescs = reinterpret_cast<const D3D12_RAYTRACING_GEOMETRY_DESC*>(pParams->pDesc->inputs.pGeometryDescs);
```

We can't do that because `sizeof(D3D12_RAYTRACING_GEOMETRY_DESC) != sizeof(NVAPI_D3D12_RAYTRACING_GEOMETRY_DESC_EX)` and the array we pass to D3D `BuildRaytracingAccelerationStructure` must have elements of size `sizeof(D3D12_RAYTRACING_GEOMETRY_DESC)`.

```cpp
if (pParams->pDesc->inputs.pGeometryDescs[i].type == NVAPI_D3D12_RAYTRACING_GEOMETRY_TYPE_OMM_TRIANGLES_EX)
```

We also shouldn't dereference `pGeometryDescs` without checking if `descLayout` is `ELEMENTS_LAYOUT_ARRAY` first.

Reply:
As discussed on chat, I had a few wrong assumptions and mistakes. Considering this and together with the response on LGD it makes perfectly sense to keep things like they are.